### ISSUE TITLE
Updating to Gradle 6.7 or above causes compile error in `buildSrc`

### ISSUE THREAD

**Issue by msridhar** [wala/WALA#892]
If I run `./gradlew wrapper --gradle-version 7.0.2` and then `./gradlew --dry-run`, I get errors when compiling `buildSrc`:

```
> Task :buildSrc:compileJava FAILED
/Users/msridhar/git-repos/WALA/buildSrc/src/main/java/com/ibm/wala/gradle/EclipseJavaCompilerFactory.java:5: error: cannot find symbol
import org.gradle.api.internal.tasks.compile.JavaCompilerFactory;
                                            ^
  symbol:   class JavaCompilerFactory
  location: package org.gradle.api.internal.tasks.compile
/Users/msridhar/git-repos/WALA/buildSrc/src/main/java/com/ibm/wala/gradle/EclipseJavaCompilerFactory.java:11: error: cannot find symbol
class EclipseJavaCompilerFactory implements JavaCompilerFactory {
                                            ^
  symbol: class JavaCompilerFactory
/Users/msridhar/git-repos/WALA/buildSrc/src/main/java/com/ibm/wala/gradle/EclipseJavaCompilerToolChain.java:5: error: cannot find symbol
import org.gradle.api.internal.tasks.AbstractJavaToolChain;
                                    ^
  symbol:   class AbstractJavaToolChain
  location: package org.gradle.api.internal.tasks
/Users/msridhar/git-repos/WALA/buildSrc/src/main/java/com/ibm/wala/gradle/EclipseJavaCompilerToolChain.java:15: error: cannot find symbol
public class EclipseJavaCompilerToolChain extends AbstractJavaToolChain {
                                                  ^
  symbol: class AbstractJavaToolChain
/Users/msridhar/git-repos/WALA/buildSrc/src/main/java/com/ibm/wala/gradle/EclipseJavaCompilerFactory.java:27: error: method does not override or implement a method from a supertype
  @Override
  ^
/Users/msridhar/git-repos/WALA/buildSrc/src/main/java/com/ibm/wala/gradle/EclipseJavaCompilerToolChain.java:22: error: method does not override or implement a method from a supertype
  @Override
  ^
/Users/msridhar/git-repos/WALA/buildSrc/src/main/java/com/ibm/wala/gradle/EclipseJavaCompilerToolChain.java:27: error: method does not override or implement a method from a supertype
  @Override
  ^
7 errors
```

We are likely blocked on upgrading Gradle until we can fix these.  @liblit could you take a look when you have some time?

Update: in further searching, I found that the relevant changes were introduced in Gradle 6.7, so we are also blocked from updating to more recent Gradle 6.x releases.

**Comment by liblit** [comment]
Will do.

**Comment by liblit** [comment]
Unfortunately, I don't think that this will be possible until gradle/gradle#15942 is fixed. I tried sneaking around Gradle 7's Java compiler limitations in various ways but got nowhere. Even if we were willing to have the ECJ builds run completely separately from regular builds (as we used to do), I don't see any way to convince Gradle 7 to use ECJ instead of a JDK-installation-based compiler.

We may have to remove our whole ECJ infrastructure, and give up on the corresponding extra checks, until gradle/gradle#15942 makes this possible again. :disappointed: I could certainly do that if we decide that it's the best way forward. Or we could linger on Gradle 6.x for a while longer to see if anything happens with gradle/gradle#15942.

**Comment by msridhar** [comment]
Thanks for the investigation, @liblit.  I am not in an extreme hurry to update to Gradle 7, though I would like to do it eventually.  I'm not too optimistic about quick progress gradle/gradle#15942, given it's been open since January.  Still, we can wait for a bit and see what happens.  If there is no progress, and we need to remove the ECJ infrastructure, we could do at least some of the checking using Error Prone, which integrates with OpenJDK's `javac` and is well supported in Gradle.  Since the code already passes the ECJ checks, swapping in an "equivalent" config of Error Prone may not be too difficult.

**Comment by liblit** [comment]
I wrote:

> I don't see any way to convince Gradle 7 to use ECJ instead of a JDK-installation-based compiler.

I believe that I have figured out how to make this work. I have a simple, stand-alone Java project that builds using Gradle 7.2 and a standard JVM 1.8 installation, but which compiles Java by running an external `ecj` command. This use of an external command can be configured per task, which is exactly what we want to allow both standard and ECJ tasks in the same Gradle invocation.

I need to do some more work to incorporate this solution into WALA’s build system. But at least I now believe that it can be done, which is a big improvement.
